editors/code/src/command/workspace-folder-formatting.ts

```diff
@@ -0,0 +1,228 @@
+import * as cp from "child_process";
+import * as vscode from "vscode";
+
+import { Cmd, Ctx } from "../context";
+import * as output from "../output";
+
+/**
+ * Format a workspace folder
+ *
+ * # Workspace folder selection
+ *
+ * - If 0 workspace folders are open, errors
+ * - If 1 workspace folder is open, automatically uses it
+ * - If >1 workspace folders are open, asks the user to choose
+ *
+ * # Tab closing
+ *
+ * Because we use the embedded Air CLI to perform the formatting, we force all
+ * relevant tabs to be saved and closed before proceeding. This has two main
+ * benefits:
+ *
+ * - The Air CLI can't know about any in-memory changes that may have occurred
+ *   in "dirty" editors. Forcing them to save and close causes any changes
+ *   to be written to disk first, syncing VS Code with the CLI.
+ *
+ * - The LSP server part of Air needs to perfectly track changes for any open
```

Comment:
I'm still not sure about that. I think I'd prefer to keep them open for now as it gives a better UX, and only close them if it turns out to be needed. As you noted the IDE doesn't generally need to close files when the whole repo is changed from under its feet via git, so I wouldn't expect problems here.

Reply:
Okay, we now save dirty open text documents that live under the workspace root, but we no longer close them

This does feel like it aligns with how a git branch switch works, so 🤞